### PR TITLE
Added support for more engines.

### PR DESCRIPTION
**Pull request by zcutlerf**
Added support for newer engines in the Instruct series and Codex, so that these models can be quickly referenced via the enum. Updated enum case descriptions to reflect this.

### CODE REVIEW

**Review by mattt — APPROVED** [review]

**Comment by mattt** [comment]
Thanks so much for these updates, @zcutlerf! Would you be able to update [the corresponding tests](https://github.com/mattt/OpenAI/blob/e5fbde9b3c50e41141054154404f1e6ed8760de4/Tests/OpenAITests/OpenAITests.swift#L26-L41) as well?

File: Sources/OpenAI/Supporting Types/Engine.swift [hunk 1]
```diff
@@ -28,44 +63,63 @@ public struct Engine: Hashable, Identifiable, Codable {
          - Note: Any task performed by a faster model like Ada
                  can be performed by a more powerful model like Curie or Davinci
          */
-        case ada
-
+        
+        
+        //Older versions of our GPT-3 models are available as davinci, curie, babbage, and ada. These are meant to be used with our fine-tuning, search, classification, and question answering endpoints.
```
Comment (mattt) [inline Sources/OpenAI/Supporting Types/Engine.swift:68]:
```suggestion
        // Older versions of our GPT-3 models are available as davinci, curie, babbage, and ada. These are meant to be used with our fine-tuning, search, classification, and question answering endpoints.
```

**Comment by zcutlerf** [comment]
Yes, I'll update them! Should the README also be changed in some way, since it lists the instruct models in a different area?

**Comment by mattt** [comment]
> Should the README also be changed in some way, since it lists the instruct models in a different area?

@zcutlerf That's a good question. It looks like OpenAI seems to be moving away from the "Instruct" distinction, and just saying that's the "new GPT-3". So I'd say:

1. Remove the "Instruct Series" section in its entirety
2. Replace the `"davinci-codex"` string literal with the enumeration case `.codeDavinci002`
3. (Maybe) Replace the older `.davinci` references with the new engines. I'm not completely sold on this, since the new scheme suggests a treadmill of keeping up with newer releases and an unclear contract for how long older engines are supported.

**Comment by mattt** [comment]
@zcutlerf After looking at a fresh response from `GET https://api.openai.com/v1/engines`, I wonder if the original design of enumeration cases makes sense anymore. Originally, there were just the 4 (Ada, Babbage, Curie, Davinci). But now, it's like... what, 50? With all sorts of weird variants and versions.

What do you think about effectively freezing this API and encouraging the use of string names going forward?

<details>
<summary><code>GET https://api.openai.com/v1/engines</code></summary>

```json
{
  "object": "list",
  "data": [
    {
      "object": "engine",
      "id": "ada",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "ada-code-search-code",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "ada-code-search-text",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "ada-search-document",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "ada-search-query",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "ada-similarity",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "babbage",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "babbage-code-search-code",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "babbage-code-search-text",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "babbage-search-document",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "babbage-search-query",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "babbage-similarity",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-cushman-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-davinci-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-davinci-002",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-davinci-edit-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-search-ada-code-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-search-ada-text-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-search-babbage-code-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "code-search-babbage-text-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "curie",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "curie-instruct-beta",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "curie-search-document",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "curie-search-query",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "curie-similarity",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "davinci",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "davinci-instruct-beta",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "davinci-search-document",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "davinci-search-query",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "davinci-similarity",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-ada-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-babbage-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-curie-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-davinci-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-davinci-002",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-davinci-edit-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-davinci-insert-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-davinci-insert-002",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-ada-doc-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-ada-query-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-babbage-doc-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-babbage-query-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-curie-doc-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-curie-query-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-davinci-doc-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-search-davinci-query-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-similarity-ada-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-similarity-babbage-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-similarity-curie-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    },
    {
      "object": "engine",
      "id": "text-similarity-davinci-001",
      "ready": true,
      "owner": "openai",
      "created": null,
      "permissions": null,
      "replicas": null,
      "max_replicas": null
    }
  ]
}
```

**Comment by zcutlerf** [comment]
I think enumeration makes it much easier to avoid string mistakes, so I'd prefer to keep it in some form. Maybe we could list the 4 they provide as "new GPT-3" models (the text-model-00n ones), since they seem to be the most supported and used ones, then the README could list a model in the enum and an older, so the usage is obvious. What do you think?

**Comment by mattt** [comment]
@zcutlerf I agree that enumeration cases have better ergonomics and correctness guarantees. But my concern is that we'll always have an incomplete list, which will inevitably force API consumers to mix-and-match string literals.

What do you think about this as a compromise — we explicitly document the best practice of extending `Engine` to define constants for engines used in your own project. For example:

```swift
extension Engine.ID {
    static var codex: Self = "code-davinci-002"
}
```

(This pattern also works for teams with access to non-public models)

**Comment by zcutlerf** [comment]
That sounds great! Does this mean we should get rid of the enum entirely, and replace all Engine types as Strings? Do we need to handle potential errors in some way, or do we just let the API throw the error?

**Comment by zcutlerf** [comment]
Ok I removed the file and fixed every error, though I'm not sure if there needs to be more scaffolding to support these choices. Let me know what you think.

**Comment by mattt** [comment]
> That sounds great! Does this mean we should get rid of the enum entirely, and replace all Engine types as Strings? Do we need to handle potential errors in some way, or do we just let the API throw the error?

Oh, sorry — to clarify, I propose that we keep `Engine.ID` and document the pattern of defining constants for engines used in a project. The `static var codex: Self = "code-davinci-002"` takes advantage of a Swift feature where types conforming to [`ExpressibleByStringLiteral`](https://developer.apple.com/documentation/swift/expressiblebystringliteral) can be initialized with a string literal, without being a string (the example is shorthand for `= .other("code-davinci-002")`). I'd be happy to take things from here. Thanks so much for your work on this and for talking through everything, @zcutlerf 😃 

Curious to hear more about anything you're building with the OpenAI APIs!

**Comment by mattt** [comment]
@zcutlerf I just opened https://github.com/mattt/OpenAI/pull/11, which implements the changes I described in https://github.com/mattt/OpenAI/pull/9#issuecomment-1117913832. Please let me know what you think. Thanks again for your contribution!

**Pull request closed without merging**